SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp

```diff
@@ -0,0 +1,874 @@
+
+  unsigned short iZone, jZone, iSol, UpdateMesh;
+  unsigned long ExtIter = 0;
+  bool DeformMesh       = false;
+
+  AD::Reset();
+
+  /*--- Prepare for recording by resetting the flow solution to the initial converged solution---*/
+
+  for(iZone = 0; iZone < nZone; iZone++) {
+
+    for (iSol=0; iSol < MAX_SOLS; iSol++){
+      if (solver_container[iZone][INST_0][MESH_0][iSol] != NULL) {
+        if (solver_container[iZone][INST_0][MESH_0][iSol]->GetAdjoint()) {
+          solver_container[iZone][INST_0][MESH_0][iSol]->SetRecording(geometry_container[iZone][INST_0][MESH_0],
+                                                                      config_container[iZone]);
+        }
+      }
+    }
+  }
+
+  /*---Enable recording and register input of the flow iteration (conservative variables or node coordinates) --- */
+
+  if(kind_recording != NONE) {
+
+    if (rank == MASTER_NODE && kind_recording == FLOW_CONS_VARS) {
+      cout << endl << "-------------------------------------------------------------------------" << endl;
+      cout << "Storing computational graph." << endl;
+    }
+
+    AD::StartRecording();
+
+    AD::Push_TapePosition();
+
+    for (iZone = 0; iZone < nZone; iZone++) {
+
+      iteration_container[iZone][INST_0]->RegisterInput(solver_container, geometry_container,
+                                                        config_container, iZone, INST_0, kind_recording);
+    }
+  }
+
+  AD::Push_TapePosition();
+
+  for (iZone = 0; iZone < nZone; iZone++) {
+
+    iteration_container[iZone][INST_0]->SetDependencies(solver_container, geometry_container, numerics_container,
+                                                        config_container, iZone, INST_0, kind_recording);
+  }
+
+  AD::Push_TapePosition();
+
+  /*--- Extract the objective function and store it --- */
+
+  SetObjFunction(kind_recording);
+
+  AD::Push_TapePosition();
+
+  if (tape_type != OBJECTIVE_FUNCTION_TAPE) {
+
+    /*--- We do the communication here to not derive wrt updated boundary data. ---*/
+
+    for(iZone = 0; iZone < nZone; iZone++) {
+
+      /*--- In principle, the mesh does not need to be updated ---*/
+      UpdateMesh = 0;
+
+      /*--- Transfer from all the remaining zones ---*/
+      for (jZone = 0; jZone < nZone; jZone++){
+        /*--- The target zone is iZone ---*/
+        if (jZone != iZone && interface_container[iZone][jZone] != NULL){
+          DeformMesh = Transfer_Data(jZone, iZone);
+          if (DeformMesh) UpdateMesh+=1;
+        }
+      }
+      /*--- If a mesh update is required due to the transfer of data ---*/
+      if (UpdateMesh > 0) DynamicMeshUpdate(iZone, ExtIter);
+    }
+
+    AD::Push_TapePosition();
+
+    for(iZone = 0; iZone < nZone; iZone++) {
+
+      AD::Push_TapePosition();
+
+      if (ZONE_SPECIFIC_TAPE && (iZone == record_zone)) {
+        DirectIteration(iZone, kind_recording);
+      }
+      else {
+        DirectIteration(iZone, kind_recording);
```

Comment:
It looks like the "if" and the "else" do the same thing?

Reply:
Sure. I guess I'll make up a small plan of everything that needs to be done (especially as there are now even two new discrete adjoint solvers, one has to do all the adaptions for those as well) and then we go through step by step. I wanted to send a mail regarding the theory anyway soon.

Thanks for finding the logic error. I'm sure I had fixed that somewhere already .. maybe on the HPC repository clone. Some stuff in between gets lost sometimes :p